File: 0011-container-with-most-water/0011-container-with-most-water.rb
# @param {Integer[]} height
# @return {Integer}
def max_area(height)
    res=0
    dif=height.length-1
    i=0
    while i<dif
            res=[res,[height[i],height[dif]].min*(dif-i)].max
        if height[i]<height[dif]
            i+=1
        else
            dif-=1
        end
    end
    res
end

File: 0011-container-with-most-water/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    let max = 0;
    const distance = new Map([[height[0], 0]])
    const points = [height[0]]

    for(let i=0;i<height.length;i++){

        for(let point of points){
            const newMax = (i-distance.get(point))*Math.min(height[i], point);
            if(newMax>max)
                max=newMax
        }

        if(height[i] > points[points.length-1]){
            points.push(height[i])
            distance.set(height[i], i)
        }
    }
    return max
};